File: src/services/TransactionService.ts
import { pool } from '../config/database';
import { createError } from '../middleware/errorHandler';

interface TransactionFilters {
  category?: string;
  startDate?: string;
  endDate?: string;
  type?: string;
}

interface TransactionData {
  category_id: string;
  amount: number;
  description: string;
  transaction_date?: string;
  merchant_name?: string;
  vendor_name?: string;
  currency?: string;
  transaction_type?: string;
}

export class TransactionService {
  async getTransactions(
    userId: string,
    page: number = 1,
    limit: number = 20,
    filters: TransactionFilters = {}
  ) {
    try {
      const offset = (page - 1) * limit;
      let whereConditions = ['t.user_id = $1'];
      let queryParams: any[] = [userId];
      let paramIndex = 2;

      // Add filters
      if (filters.category) {
        whereConditions.push(`c.name ILIKE $${paramIndex}`);
        queryParams.push(`%${filters.category}%`);
        paramIndex++;
      }

      if (filters.startDate) {
        whereConditions.push(`t.transaction_date >= $${paramIndex}`);
        queryParams.push(filters.startDate);
        paramIndex++;
      }

      if (filters.endDate) {
        whereConditions.push(`t.transaction_date <= $${paramIndex}`);
        queryParams.push(filters.endDate);
        paramIndex++;
      }

      if (filters.type) {
        whereConditions.push(`t.transaction_type = $${paramIndex}`);
        queryParams.push(filters.type);
        paramIndex++;
      }

      const whereClause = whereConditions.join(' AND ');

      // Get total count
      const countQuery = `
        SELECT COUNT(*)
        FROM transactions t
        LEFT JOIN categories c ON t.category_id = c.id
        WHERE ${whereClause}
      `;
      const countResult = await pool.query(countQuery, queryParams.slice(0, paramIndex - 1));
      const total = parseInt(countResult.rows[0].count);

      // Get transactions
      const query = `
        SELECT 
          t.*,
          c.name as category_name,
          c.color as category_color,
          c.icon as category_icon
        FROM transactions t
        LEFT JOIN categories c ON t.category_id = c.id
        WHERE ${whereClause}
        ORDER BY t.transaction_date DESC, t.created_at DESC
        LIMIT $${paramIndex} OFFSET $${paramIndex + 1}
      `;

      queryParams.push(limit, offset);
      const result = await pool.query(query, queryParams);

      return {
        transactions: result.rows,
        total,
        page,
        totalPages: Math.ceil(total / limit)
      };
    } catch (error) {
      throw createError('Failed to retrieve transactions', 500);
    }
  }

  async createTransaction(userId: string, data: TransactionData) {
    try {
      const query = `
        INSERT INTO transactions (
          user_id, category_id, amount, description, transaction_date,
          vendor_name, currency, transaction_type
        )
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
        RETURNING *
      `;

      const vendorName = data.vendor_name || data.merchant_name || 'Unknown Vendor';
      const currency = data.currency || 'USD';

      const result = await pool.query(query, [
        userId,
        data.category_id,
        data.amount,
        data.description,
        data.transaction_date || new Date().toISOString().split('T')[0],
        vendorName,
        currency,
        data.transaction_type || 'expense'
      ]);

      return result.rows[0];
    } catch (error) {
      throw createError('Failed to create transaction', 500);
    }
  }

  async getTransactionById(transactionId: string, userId: string) {
    const client = await pool.connect();
    try {
      console.log('🔍 [Transaction] Fetching transaction details for:', { transactionId, userId });
      
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      const query = `
        SELECT 
          t.*,
          c.name as category_name,
          c.color as category_color,
          c.icon as category_icon,
          rp.extracted_data as receipt_data,
          rp.id as receipt_id
        FROM transactions t
        LEFT JOIN categories c ON t.category_id = c.id
        LEFT JOIN receipt_processing rp ON rp.transaction_id = t.id
        WHERE t.id = $1 AND t.user_id = $2
      `;

      const result = await client.query(query, [transactionId, userId]);

      if (result.rows.length === 0) {
        throw createError('Transaction not found', 404);
      }

      const transaction = result.rows[0];
      
      console.log('🔍 [Transaction] Query result:', {
        transactionId: transaction.id,
        hasReceiptData: !!transaction.receipt_data,
        receiptId: transaction.receipt_id,
        receiptDataLength: transaction.receipt_data ? transaction.receipt_data.length : 0
      });
      
      // Handle receipt data (already parsed as JSONB from PostgreSQL)
      if (transaction.receipt_data) {
        try {
          // If it's already an object, use it directly; if it's a string, parse it
          if (typeof transaction.receipt_data === 'string') {
            transaction.receipt_details = JSON.parse(transaction.receipt_data);
          } else {
            transaction.receipt_details = transaction.receipt_data;
          }
          
          console.log('✅ [Transaction] Receipt data processed successfully');
          console.log('🔍 [Transaction] Receipt data structure:', {
            hasExtractedText: !!transaction.receipt_details.extractedText,
            hasParsedData: !!transaction.receipt_details.parsedData,
            itemsCount: transaction.receipt_details.parsedData?.items?.length || 0,
            dataType: typeof transaction.receipt_data
          });
        } catch (error) {
          console.error('❌ [Transaction] Error processing receipt data:', error);
          transaction.receipt_details = null;
        }
      } else {
        console.log('⚠️ [Transaction] No receipt data found for transaction');
      }

      return transaction;
    } catch (error) {
      console.error('❌ [Transaction] Failed to retrieve transaction:', error);
      throw createError('Failed to retrieve transaction', 500);
    } finally {
      client.release();
    }
  }

  async updateTransaction(transactionId: string, userId: string, data: Partial<TransactionData>) {
    try {
      const setClauses = [];
      const queryParams = [];
      let paramIndex = 1;

      Object.entries(data).forEach(([key, value]) => {
        if (value !== undefined) {
          setClauses.push(`${key} = $${paramIndex}`);
          queryParams.push(value);
          paramIndex++;
        }
      });

      if (setClauses.length === 0) {
        throw createError('No data provided for update', 400);
      }

      setClauses.push(`updated_at = CURRENT_TIMESTAMP`);
      queryParams.push(transactionId, userId);

      const query = `
        UPDATE transactions
        SET ${setClauses.join(', ')}
        WHERE id = $${paramIndex} AND user_id = $${paramIndex + 1}
        RETURNING *
      `;

      const result = await pool.query(query, queryParams);

      if (result.rows.length === 0) {
        throw createError('Transaction not found', 404);
      }

      return result.rows[0];
    } catch (error) {
      throw createError('Failed to update transaction', 500);
    }
  }

  async deleteTransaction(transactionId: string, userId: string) {
    const client = await pool.connect();
    try {
      // Set user context for Row Level Security
      await client.query(`SET app.current_user_id = '${userId}'`);
      
      const query = `
        DELETE FROM transactions
        WHERE id = $1 AND user_id = $2
        RETURNING id
      `;

      const result = await client.query(query, [transactionId, userId]);

      if (result.rows.length === 0) {
        throw createError('Transaction not found', 404);
      }
    } catch (error) {
      console.error('Error deleting transaction:', error);
      throw createError('Failed to delete transaction', 500);
    } finally {
      client.release();
    }
  }

  async getSpendingSummary(userId: string, period: string = 'month') {
    try {
      let dateCondition = '';
      
      switch (period) {
        case 'week':
          dateCondition = `AND transaction_date >= date_trunc('week', CURRENT_DATE)`;
          break;
        case 'month':
          dateCondition = `AND transaction_date >= date_trunc('month', CURRENT_DATE)`;
          break;
        case 'year':
          dateCondition = `AND transaction_date >= date_trunc('year', CURRENT_DATE)`;
          break;
        default:
          dateCondition = `AND transaction_date >= date_trunc('month', CURRENT_DATE)`;
      }

      const query = `
        SELECT
          COALESCE(SUM(CASE WHEN transaction_type = 'income' THEN amount ELSE 0 END), 0) as total_income,
          COALESCE(SUM(CASE WHEN transaction_type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
          COUNT(*) as total_transactions,
          COUNT(CASE WHEN transaction_type = 'income' THEN 1 END) as income_count,
          COUNT(CASE WHEN transaction_type = 'expense' THEN 1 END) as expense_count
        FROM transactions
        WHERE user_id = $1 ${dateCondition}
      `;

      const result = await pool.query(query, [userId]);
      const summary = result.rows[0];

      return {
        period,
        totalIncome: parseFloat(summary.total_income),
        totalExpenses: parseFloat(summary.total_expenses),
        netAmount: parseFloat(summary.total_income) - parseFloat(summary.total_expenses),
        totalTransactions: parseInt(summary.total_transactions),
        incomeCount: parseInt(summary.income_count),
        expenseCount: parseInt(summary.expense_count)
      };
    } catch (error) {
      throw createError('Failed to get spending summary', 500);
    }
  }

  async getCategorySummary(userId: string, period: string = 'month') {
    try {
      let dateCondition = '';
      
      switch (period) {
        case 'week':
          dateCondition = `AND t.transaction_date >= date_trunc('week', CURRENT_DATE)`;
          break;
        case 'month':
          dateCondition = `AND t.transaction_date >= date_trunc('month', CURRENT_DATE)`;
          break;
        case 'year':
          dateCondition = `AND t.transaction_date >= date_trunc('year', CURRENT_DATE)`;
          break;
        default:
          dateCondition = `AND t.transaction_date >= date_trunc('month', CURRENT_DATE)`;
      }

      const query = `
        SELECT
          c.id,
          c.name,
          c.color,
          c.icon,
          COALESCE(SUM(CASE WHEN t.transaction_type = 'expense' THEN t.amount ELSE 0 END), 0) as total_spent,
          COALESCE(SUM(CASE WHEN t.transaction_type = 'income' THEN t.amount ELSE 0 END), 0) as total_income,
          COUNT(t.id) as transaction_count
        FROM categories c
        LEFT JOIN transactions t ON c.id = t.category_id ${dateCondition}
        WHERE c.user_id = $1
        GROUP BY c.id, c.name, c.color, c.icon
        ORDER BY total_spent DESC
      `;

      const result = await pool.query(query, [userId]);

      return {
        period,
        categories: result.rows.map((row: any) => ({
          id: row.id,
          name: row.name,
          color: row.color,
          icon: row.icon,
          totalSpent: parseFloat(row.total_spent),
          totalIncome: parseFloat(row.total_income),
          transactionCount: parseInt(row.transaction_count),
          percentage: 0 // Will be calculated based on total spending
        }))
      };
    } catch (error) {
      throw createError('Failed to get category summary', 500);
    }
  }

  async getCategoryIdByName(userId: string, categoryName: string): Promise<string | null> {
    try {
      const query = `
        SELECT id FROM categories
        WHERE user_id = $1 AND LOWER(name) = LOWER($2)
        LIMIT 1
      `;

      const result = await pool.query(query, [userId, categoryName]);
      return result.rows[0]?.id || null;
    } catch (error) {
      return null;
    }
  }
}

File: src/services/transaction_service.py
from typing import List, Dict, Any, Optional
import uuid
from datetime import datetime
from src.config.database import DatabaseManager
from src.models.transaction import TransactionCreate, TransactionUpdate, TransactionFilter
from src.utils.logger import logger


class TransactionService:
    """Service for transaction-related operations"""
    
    async def get_transactions(self, user_id: str, filters: TransactionFilter) -> Dict[str, Any]:
        """Get user transactions with filtering and pagination"""
        try:
            # Build WHERE clause
            where_conditions = ["t.user_id = $1"]
            params = [uuid.UUID(user_id)]
            param_count = 2
            
            if filters.category_id:
                where_conditions.append(f"t.category_id = ${param_count}")
                # filters.category_id is already a UUID object from the API layer
                if isinstance(filters.category_id, str):
                    params.append(uuid.UUID(filters.category_id))
                else:
                    params.append(filters.category_id)
                param_count += 1
            
            if filters.transaction_type:
                where_conditions.append(f"t.transaction_type = ${param_count}")
                params.append(filters.transaction_type)
                param_count += 1
            
            if filters.start_date:
                where_conditions.append(f"t.transaction_date >= ${param_count}")
                # filters.start_date is already a datetime object from the API layer
                if isinstance(filters.start_date, str):
                    params.append(datetime.fromisoformat(filters.start_date.replace('Z', '+00:00')))
                else:
                    params.append(filters.start_date)
                param_count += 1
            
            if filters.end_date:
                where_conditions.append(f"t.transaction_date <= ${param_count}")
                # filters.end_date is already a datetime object from the API layer
                if isinstance(filters.end_date, str):
                    params.append(datetime.fromisoformat(filters.end_date.replace('Z', '+00:00')))
                else:
                    params.append(filters.end_date)
                param_count += 1
            
            if filters.vendor_name:
                where_conditions.append(f"t.vendor_name ILIKE ${param_count}")
                params.append(f"%{filters.vendor_name}%")
                param_count += 1
            
            # Calculate offset
            offset = (filters.page - 1) * filters.limit
            
            query = f"""
                SELECT 
                    t.id, t.user_id, t.category_id, t.vendor_name, t.amount, t.tax_amount,
                    t.currency, t.transaction_date, t.description, t.receipt_url,
                    t.transaction_type, t.created_at, t.updated_at,
                    c.name as category_name
                FROM transactions t
                LEFT JOIN categories c ON t.category_id = c.id
                WHERE {' AND '.join(where_conditions)}
                ORDER BY t.transaction_date DESC
                LIMIT ${param_count} OFFSET ${param_count + 1}
            """
            
            params.extend([filters.limit, offset])
            
            async with DatabaseManager(user_id) as db:
                # Get transactions with pagination
                rows = await db.fetch(query, *params)
                transactions = [dict(row) for row in rows]
                
                # Get total count for pagination
                count_query = f"""
                    SELECT COUNT(*) as total
                    FROM transactions t
                    LEFT JOIN categories c ON t.category_id = c.id
                    WHERE {' AND '.join(where_conditions)}
                """
                count_params = params[:-2]  # Remove limit and offset params
                total_result = await db.fetchrow(count_query, *count_params)
                total_count = total_result['total'] if total_result else 0
                
                return {
                    "transactions": transactions,
                    "total": total_count,
                    "page": filters.page,
                    "limit": filters.limit,
                    "total_pages": (total_count + filters.limit - 1) // filters.limit
                }
        
        except Exception as e:
            logger.error(f"Error getting transactions: {e}")
            raise e
    
    async def create_transaction(self, user_id: str, transaction_data: TransactionCreate) -> Dict[str, Any]:
        """Create a new transaction"""
        try:
            query = """
                INSERT INTO transactions 
                (user_id, category_id, vendor_name, amount, tax_amount, currency, 
                 transaction_date, description, transaction_type)
                VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
                RETURNING id, user_id, category_id, vendor_name, amount, tax_amount,
                         currency, transaction_date, description, receipt_url,
                         transaction_type, created_at, updated_at
            """
            
            async with DatabaseManager(user_id) as db:
                transaction = await db.fetchrow(
                    query,
                    uuid.UUID(user_id),
                    transaction_data.category_id,
                    transaction_data.vendor_name,
                    transaction_data.amount,
                    transaction_data.tax_amount or 0,
                    transaction_data.currency,
                    transaction_data.transaction_date,
                    transaction_data.description,
                    transaction_data.transaction_type
                )
                
                return dict(transaction)
        
        except Exception as e:
            logger.error(f"Error creating transaction: {e}")
            raise e
    
    async def get_transaction(self, user_id: str, transaction_id: str) -> Optional[Dict[str, Any]]:
        """Get a specific transaction"""
        try:
            query = """
                SELECT 
                    t.id, t.user_id, t.category_id, t.vendor_name, t.amount, t.tax_amount,
                    t.currency, t.transaction_date, t.description, t.receipt_url,
                    t.transaction_type, t.created_at, t.updated_at,
                    c.name as category_name, c.color as category_color, c.icon as category_icon,
                    rp.id as receipt_id, rp.extracted_data
                FROM transactions t
                LEFT JOIN categories c ON t.category_id = c.id
                LEFT JOIN receipt_processing rp ON rp.transaction_id = t.id
                WHERE t.id = $1 AND t.user_id = $2
            """
            
            async with DatabaseManager(user_id) as db:
                transaction = await db.fetchrow(
                    query, 
                    uuid.UUID(transaction_id),
                    uuid.UUID(user_id)
                )
                
                if not transaction:
                    return None
                
                result = dict(transaction)
                
                # Process receipt data if it exists
                if result.get('extracted_data'):
                    try:
                        import json
                        # Parse the extracted_data JSON string
                        extracted_data = result['extracted_data']
                        if isinstance(extracted_data, str):
                            parsed_data = json.loads(extracted_data)
                        else:
                            parsed_data = extracted_data
                        
                        # Create receipt_details structure similar to frontend expectations
                        result['receipt_details'] = {
                            'extractedText': '',  # OCR text is not stored separately in current backend
                            'parsedData': parsed_data,
                            'parsed_data': parsed_data  # Also include snake_case version for compatibility
                        }
                        
                        # Clean up the raw extracted_data from the response
                        del result['extracted_data']
                        
                    except (json.JSONDecodeError, Exception) as e:
                        logger.warning(f"Failed to parse receipt extracted_data: {e}")
                        # Remove the raw extracted_data if it can't be parsed
                        del result['extracted_data']
                
                return result
        
        except Exception as e:
            logger.error(f"Error getting transaction: {e}")
            raise e
    
    async def update_transaction(
        self, 
        user_id: str, 
        transaction_id: str, 
        transaction_data: TransactionUpdate
    ) -> Optional[Dict[str, Any]]:
        """Update a transaction"""
        try:
            # Build dynamic update query
            update_fields = []
            params = []
            param_count = 1
            
            if transaction_data.vendor_name is not None:
                update_fields.append(f"vendor_name = ${param_count}")
                params.append(transaction_data.vendor_name)
                param_count += 1
            
            if transaction_data.amount is not None:
                update_fields.append(f"amount = ${param_count}")
                params.append(transaction_data.amount)
                param_count += 1
            
            if transaction_data.tax_amount is not None:
                update_fields.append(f"tax_amount = ${param_count}")
                params.append(transaction_data.tax_amount)
                param_count += 1
            
            if transaction_data.currency is not None:
                update_fields.append(f"currency = ${param_count}")
                params.append(transaction_data.currency)
                param_count += 1
            
            if transaction_data.transaction_date is not None:
                update_fields.append(f"transaction_date = ${param_count}")
                params.append(transaction_data.transaction_date)
                param_count += 1
            
            if transaction_data.description is not None:
                update_fields.append(f"description = ${param_count}")
                params.append(transaction_data.description)
                param_count += 1
            
            if transaction_data.category_id is not None:
                update_fields.append(f"category_id = ${param_count}")
                params.append(transaction_data.category_id)
                param_count += 1
            
            if transaction_data.transaction_type is not None:
                update_fields.append(f"transaction_type = ${param_count}")
                params.append(transaction_data.transaction_type)
                param_count += 1
            
            if not update_fields:
                # No fields to update
                return await self.get_transaction(user_id, transaction_id)
            
            # Add where clause parameters
            params.extend([uuid.UUID(transaction_id), uuid.UUID(user_id)])
            
            query = f"""
                UPDATE transactions 
                SET {', '.join(update_fields)}
                WHERE id = ${param_count} AND user_id = ${param_count + 1}
                RETURNING id, user_id, category_id, vendor_name, amount, tax_amount,
                         currency, transaction_date, description, receipt_url,
                         transaction_type, created_at, updated_at
            """
            
            async with DatabaseManager(user_id) as db:
                transaction = await db.fetchrow(query, *params)
                return dict(transaction) if transaction else None
        
        except Exception as e:
            logger.error(f"Error updating transaction: {e}")
            raise e
    
    async def delete_transaction(self, user_id: str, transaction_id: str) -> bool:
        """Delete a transaction"""
        try:
            query = "DELETE FROM transactions WHERE id = $1 AND user_id = $2"
            
            async with DatabaseManager(user_id) as db:
                result = await db.execute(
                    query,
                    uuid.UUID(transaction_id),
                    uuid.UUID(user_id)
                )
                
                # Check if any row was affected
                return result.split()[-1] == "1"
        
        except Exception as e:
            logger.error(f"Error deleting transaction: {e}")
            raise e
    
    async def get_spending_summary(
        self, 
        user_id: str, 
        start_date: Optional[str] = None, 
        end_date: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get spending summary for a period"""
        try:
            where_conditions = ["user_id = $1"]
            params = [uuid.UUID(user_id)]
            param_count = 2
            
            if start_date:
                where_conditions.append(f"transaction_date >= ${param_count}")
                if isinstance(start_date, str):
                    params.append(datetime.fromisoformat(start_date.replace('Z', '+00:00')))
                else:
                    params.append(start_date)
                param_count += 1
            
            if end_date:
                where_conditions.append(f"transaction_date <= ${param_count}")
                if isinstance(end_date, str):
                    params.append(datetime.fromisoformat(end_date.replace('Z', '+00:00')))
                else:
                    params.append(end_date)
                param_count += 1
            
            query = f"""
                SELECT 
                    COALESCE(SUM(CASE WHEN transaction_type = 'income' THEN amount ELSE 0 END), 0) as total_income,
                    COALESCE(SUM(CASE WHEN transaction_type = 'expense' THEN amount ELSE 0 END), 0) as total_expenses,
                    COUNT(*) as transaction_count
                FROM transactions
                WHERE {' AND '.join(where_conditions)}
            """
            
            async with DatabaseManager(user_id) as db:
                row = await db.fetchrow(query, *params)
                
                return {
                    "total_income": float(row["total_income"]),
                    "total_expenses": float(row["total_expenses"]),
                    "net_amount": float(row["total_income"]) - float(row["total_expenses"]),
                    "transaction_count": row["transaction_count"],
                    "period": f"{start_date or 'all time'} to {end_date or 'now'}"
                }
        
        except Exception as e:
            logger.error(f"Error getting spending summary: {e}")
            raise e
    
    async def get_category_summary(
        self, 
        user_id: str, 
        start_date: Optional[str] = None, 
        end_date: Optional[str] = None
    ) -> List[Dict[str, Any]]:
        """Get category breakdown for a period"""
        try:
            where_conditions = ["t.user_id = $1"]
            params = [uuid.UUID(user_id)]
            param_count = 2
            
            if start_date:
                where_conditions.append(f"t.transaction_date >= ${param_count}")
                params.append(datetime.fromisoformat(start_date.replace('Z', '+00:00')))
                param_count += 1
            
            if end_date:
                where_conditions.append(f"t.transaction_date <= ${param_count}")
                params.append(datetime.fromisoformat(end_date.replace('Z', '+00:00')))
                param_count += 1
            
            query = f"""
                SELECT 
                    t.category_id,
                    c.name as category_name,
                    c.color as category_color,
                    c.icon as category_icon,
                    SUM(t.amount) as total_amount,
                    COUNT(*) as transaction_count,
                    (SUM(t.amount) * 100.0 / NULLIF(SUM(SUM(t.amount)) OVER (), 0)) as percentage
                FROM transactions t
                LEFT JOIN categories c ON t.category_id = c.id
                WHERE {' AND '.join(where_conditions)} AND t.transaction_type = 'expense'
                GROUP BY t.category_id, c.name, c.color, c.icon
                ORDER BY total_amount DESC
            """
            
            async with DatabaseManager(user_id) as db:
                rows = await db.fetch(query, *params)
                
                return [
                    {
                        "category_id": row["category_id"],
                        "category_name": row["category_name"] or "Unknown",
                        "category_color": row["category_color"] or "#6B7280",
                        "category_icon": row["category_icon"] or "📁",
                        "total_amount": float(row["total_amount"]),
                        "transaction_count": row["transaction_count"],
                        "percentage": float(row["percentage"] or 0)
                    }
                    for row in rows
                ]
        
        except Exception as e:
            logger.error(f"Error getting category summary: {e}")
            raise e